Got Ansible own modules? Run them from within the Salt just like that!

isbm opened the pull request:
### What does this PR do?

In case you want to migrate away from Ansible, but have already modules written for it — no worries. Just toss them in and run them from Salt and Salt states.

### Limitations

Currently, only Python-written modules are supported. Support for other languages are still in progress.

### Requirements

The entire Ansible should be installed on the Minion. No remote execution at the moment. Continuation of this work can pull required files from the Ansible through the Salt transport. These files should be installed on Master only (in a future) and then thus get executed on the Minions.

### Description and Usage

Normally Salt modules are usually `module.function`. That is, a typical call of a function from the module will be: `salt hostname module.function some=parameters`. But since we are connecting the entire Ansible world to the Salt universe, we would like to keep the namespace as this is inside the Ansible. That said, any Ansible module is placed in the sub-directory that potentially can clash with other same-named module. Regardless the future plans of Ansible community, we already have a namespace introduced. Therefore any Ansible module is called with a full "path", prefixed with `ansible`:

    salt \* ansible.system.ping data="Hello from Ansible via Salt"

If you use typical development environment, this will result to the following:
```
saltdevel:
    ----------
    ping:
        Hello from Ansible via Salt
```

### Listing available Ansible modules

In order to list all available Ansible modules, call `ansible.list`:

    salt yourminion ansible.list

To filter them out, just type a string that matches inside them, e.g. "linux":
```
$ salt saltdevel ansible.list info
saltdevel:
    - net_tools.ipinfoio_facts
    - network.cloudengine.ce_info_center_debug
    - network.cloudengine.ce_info_center_global
    - network.cloudengine.ce_info_center_log
    - network.cloudengine.ce_info_center_trap
```

### Getting more information about the module

You can read all the details about any Ansible modules by issuing `ansible.help`. For example:
```
$ salt saltdevel ansible.help net_tools.ipinfoio_facts
saltdevel:
    ----------
    Available sections on module "net_tools.ipinfoio_facts":
        - version_added
        - author
        - short_description
        - notes
        - options
        - module
    Description:
        - Gather IP geolocation facts of a host's IP address using ipinfo.io API
```
At this point you may ask for `short_description`, `notes` etc. For example, to know how to use it, call `options`:
```
$ salt saltdevel ansible.help net_tools.ipinfoio_facts options
saltdevel:
    ----------
    options:
        ----------
        http_agent:
            ----------
            default:
                ansible-ipinfoio-module/0.0.1
            description:
                - Set http user agent
            required:
                False
        timeout:
            ----------
            default:
                10
            description:
                - HTTP connection timeout in seconds
            required:
                False
```

### Calling Ansible modules

Calling Ansible modules are just like any other Salt modules, straight-forward, except they will be needing to stay within their own namespace, which is different than we are using typically in Salt:
```
$ salt saltdevel ansible.net_tools.ipinfoio_facts                                    
saltdevel:
    ----------
    ansible_facts:
        ----------
        city:
            Nuremberg
        country:
            DE
        hostname:
            charybdis-ext.suse.de
        ip:
            195.135.221.2
        loc:
            49.4478,11.0683
        org:
            AS29298 SUSE LINUX GmbH
        postal:
            90409
        region:
            Bavaria
    changed:
        False
```

### Calling from the SLS files

That is also supported. Naturally, you can use typical `module.run` if you like, but you've also got a dedicated Ansible state module. The syntax/example is the same as for new syntax for `module.run`. It also supports a "batch mode", or "Ansible playbook mode" where tasks can be executed from within one SLS file:
```yaml
task_id:        # ID of your task
  ansible.call:  # Mandatory. This calls the modules
    - namespace.module:
      - params
    - namespace.another.module:
      - params
      - some: keywords
```

For example, this state will fail, because `foo.bar` module does not exist, as well as `network.junos.junos_package` requires some parameters. However it will still call everything one after another:

```yaml
run_test_echo:
  ansible.call:
    - system.ping:
        - data: "Hello from Salt"
    - network.junos.junos_package:
    - foo.bar:
 ```
This results to the following output:
```
$ salt saltdevel state.apply ans
saltdevel:
----------
          ID: run_test_echo
    Function: ansible.call
      Result: False
     Comment: 
     Started: 15:33:04.367029
    Duration: 1409.689 ms
     Changes:   
              ----------
              foo.bar:
                  Module "foo.bar" failed. Error message: (KeyError) 'ansible.foo.bar'
              net_tools.ipinfoio_facts:
                  ----------
              network.junos.junos_package:
                  ----------
                  failed:
                      True
                  msg:
                      missing required arguments: src
              system.ping:
                  ----------
                  ping:
                      Hello from Salt

Summary for saltdevel
------------
Succeeded: 0 (changed=1)
Failed:    1
------------
Total states run:     1
Total run time:   1.410 s
```

### Tests written?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

Commit:
Initial start of ansible module

Commit:
Add initial module resolver, implement modules map

Commit:
Add help resolver on a particular Python module of Ansible.

Commit:
Improve error handling for the loaded module

Commit:
Implement module lister

Commit:
Set dynamically all possible Ansible callers as functions

Commit:
Allow long names with dots in between

Commit:
Add missed import

Commit:
Prepare arguments for the module

Commit:
Register the full name of the module

Commit:
Implement basic module caller (Python modules only)

Commit:
Catch loader error as well

Commit:
Call description from help by default

Commit:
Prevent dummy modules to be executed to nothing

Commit:
Add quotes to the help command

Commit:
Initial scaffolding for Ansible state module

Commit:
Take to the account popen's STDERR as well.

Commit:
Wrap caller into a class

Commit:
Implement basic Ansible state call

Commit:
Handle errors

Commit:
Remove dead code

Commit:
Mimic the direct call of the module to reuse internal arg parsing

Commit:
Add the license header

Commit:
Add scaffold for the unit test

Commit:
Add unit test for documentation extractor

Commit:
Do not use mock but real resolver class

Commit:
Add unit test for modules filtering

Commit:
Refactor to setup/teardown methods

Commit:
Add unit test resolver module loader failure

Commit:
Improve module loader failure

Commit:
Add unit test for module loader

Commit:
Add unit test for module loader failure

Commit:
Don't bother with non-pytest unit tests

isbm commented:
@meaksh @dincamihai @cachedout — guys, Merry Christmas. :wink: Keep an eye on it (ignore lints and little bits, let it boil, I will look after a while).
@thatch45 Just like promised. :sunglasses:

Commit:
Lintfix: W1699

Commit:
Lintfix: E0102

Commit:
Lintfix: E9502

Commit:
Lintfix: W0108

Commit:
Lintfix: W0611

Commit:
Lintfix: E8701 (PEP8 E701), C0321

Commit:
Lintfix: W0602

Commit:
Fix TypeError for argument of type 'LoaderError' is not iterable

cachedout commented:
re-run lint

cachedout commented:
This is awesome!

My main concern here is the departure from the `salt_module.salt_function` namespace. I think we could still maintain our current standard and just call Ansible modules with something like `ansible.call ansible_module.ansible_function`.

How do you feel about that, @isbm ?

isbm commented:
@cachedout I feel not very well about. :wink: We actually thought about that and expected this concern, still chose to depart from that for these two reasons:
— Ansible is placing everything under different directories. While it _currently_ maintains distinct names, yet their entire structure more like objects or at least namespaced structure. That said, nobody stops any user have two modules with the same name in different places. Since Ansible module does only one thing at a time, this problem might arise pretty much soon. Moreover, these modules are from kind of "another world", where keeping them isolated is good for two another reasons: a) Users mainly will reuse their Ansible modules from Salt in order not to rewrite them. They can happily place their same-name modules in the different directories for more convenient handling etc. b) We don't need to worry for name clashes. That said, at least keeping the namespace seems to be something that is rather needed.
— The implementation option that does something like `ansible.call 'some.ansible.module'` looks to me pretty messy. It is better and cleaner to do Python way, e.g. in `import foo.bar.blah` rather then go Perl/Ruby/PHP way: `require "foo.bar.blah"` where that "require" would resembled "ansible.call".

On the other hand, why not depart anyway? Modules in Salt are `module.function`, OK. But what possibly can go wrong with `module.class.function` or `module.submodule.function` anyway as long as we have a good introspection where anyone can call "help" over "class/submodule" and see what functions are there? I find this an advantage that should not be overused, sure, but having it when it is needed isn't really a problem. Especially that Ansible caller already lists you all modules with the namespaces. You can just copypaste that and it is there.

Commit:
Fix value is not iterable error

gtmanfred commented:
Personally I do not like the departing from the module.function model.  We already have other similar things that work with the whatever.call model.

Between the salt module

https://docs.saltstack.com/en/latest/ref/runners/all/salt.runners.salt.html#salt.runners.salt.cmd

And the way that most proxy minions are written.

https://github.com/saltstack/salt/blob/2017.7/salt/modules/esxi.py#L52

We already have this kind of behavior in place.

My 2¢, Thanks.

Daniel

isbm commented:
@gtmanfred It is not a runner module, and the paradigm is different. When an Ansible module is called, it is called like being a part of Salt (this is a migration feature, not just generic execution implementation), while in your examples parameters are passed to somewhere else towards _generic_ function target.

And even your example isn't the best one as it would be better to incorporate there the model I propose. Alongside of the weird path `salt.runners.salt` you have a) `cmd` and b) `execute` functions with the only difference that one should be remembered as for local stuff and the other one is exactly the same as previous, except works only on remote. It would be much cleaner to have "what.do.where <how>":
```
salt.runners.execute.local <function> <parameters>
salt.runners.execute.remote <target> <function> <parameters>
```

And yes, unfortunately we have `module.run` instead of `module.execute` to the story above... :disappointed:

isbm commented:
@thatch45 thoughts?

isbm commented:
@s0undt3ch do you have any ideas why [tests on Ubuntu 16](https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/2002/) fails that way? I am trying with PyTest and "traditional" tests in different directions and they are passing...

Commit:
Remove extra assertions

isbm commented:
Go Go Jenkins!

cachedout commented:
@isbm Right now @thatch45 is out of the office but I will get him to weigh in here as soon as possible. Thanks for your patience.

cachedout commented on salt/modules/ansiblegate.py:
```diff
@@ -0,0 +1,250 @@
+        if args:
+            kwargs['_raw_params'] = ' '.join(args)
+        js_args = '{{"ANSIBLE_MODULE_ARGS": {args}}}'.format(args=json.dumps(kwargs))
+        js_out = subprocess.Popen(["echo", "{0}".format(js_args)], stdout=subprocess.PIPE)
```
What do you think about using `salt.utils.timed_subprocess` here and in the line below so that we can bail out if we hang?

isbm replied:
Good point. Let me check for it. Getting back here ASAP.

cachedout commented on salt/modules/ansiblegate.py:
```diff
@@ -0,0 +1,250 @@
+    Set all Ansible modules callables
+    :return:
+    '''
+    def _mkf(cmd_name, doc):
```
A more descriptive function name here would help make this more maintainable in the future. :]

isbm replied:
OK

Commit:
Add more descriptive function name

Commit:
Use timed subprocess for bailing out on hanging situation

isbm commented:
@cachedout OK, this now should be good enough.

Commit:
Add timeout option, default to 20 minutes

Commit:
Add root-level documentation to the module

isbm commented:
@cachedout I've also added an option `ansible_timeout` that can be set in `/etc/salt/minion` defaulting to 20 minutes and a documentation. No idea why GitHub tells me this branch is out of date, I just rebased...

isbm commented:
@cachedout anything else I need to do here?

isbm commented:
@cachedout @thatch45 is there any chances to get this done? It is blocking further development of e.g. conversion Ansible playbooks to SLS. :cry: ...
